### ISSUE TITLE
Memory leaks in vkCreateInstance with standard validation layer enabled

### ISSUE THREAD

**Issue by polarina** [KhronosGroup/Vulkan-LoaderAndValidationLayers#1787]
The following program demonstrates memory leaks in vkCreateInstance when VK_LAYER_LUNARG_standard_validation is enabled.

Tested against official Linux build of VulkanSDK 1.0.49.0 on Debian Stretch. When `enabledLayerCount` is set to to zero in the sample program, no leaks are detected by either Valgrind or LeakSanitizer.

With Valgrind:
Compile: `gcc vkleak.c -lvulkan -g`
Run: `valgrind ./a.out`

With LeakSanitizer:
Compile: `gcc vkleak.c -lvulkan -g -fsanitize=address`
Run: `./a.out`

```
#include <vulkan/vulkan.h>

int main()
{
	VkInstance instance;
	VkResult result;

	result = vkCreateInstance(
		&(VkInstanceCreateInfo) {
			.sType = VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO,
			.enabledLayerCount = 1,
			.ppEnabledLayerNames = (const char *[]) {
				"VK_LAYER_LUNARG_standard_validation",
			},
		},
		NULL,
		&instance);

	if (result == VK_SUCCESS)
	{
		vkDestroyInstance(instance, NULL);
	}

	return result != VK_SUCCESS;
}
```

Output from Valgrind:
```
$ valgrind --leak-check=full ./a.out 
==31678== Memcheck, a memory error detector
==31678== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==31678== Using Valgrind-3.12.0.SVN and LibVEX; rerun with -h for copyright info
==31678== Command: ./a.out
==31678== 
==31678== 
==31678== HEAP SUMMARY:
==31678==     in use at exit: 176,055 bytes in 994 blocks
==31678==   total heap usage: 26,759 allocs, 25,765 frees, 80,452,011 bytes allocated
==31678== 
==31678== 456 bytes in 1 blocks are definitely lost in loss record 904 of 918
==31678==    at 0x4C2C21F: operator new(unsigned long) (vg_replace_malloc.c:334)
==31678==    by 0x1113F55B: ???
==31678==    by 0x1112968B: ???
==31678==    by 0x1148049F: ???
==31678==    by 0x117E8172: ???
==31678==    by 0x4E500AE: loader_create_instance_chain (in /home/polarina/VulkanSDK/1.0.49.0/x86_64/lib/libvulkan.so.1.0.49)
==31678==    by 0x4E52A89: vkCreateInstance (in /home/polarina/VulkanSDK/1.0.49.0/x86_64/lib/libvulkan.so.1.0.49)
==31678==    by 0x108831: main (vkleak.c:8)
==31678== 
==31678== 552 (544 direct, 8 indirect) bytes in 1 blocks are definitely lost in loss record 905 of 918
==31678==    at 0x4C2C21F: operator new(unsigned long) (vg_replace_malloc.c:334)
==31678==    by 0x1147FF5B: ???
==31678==    by 0x114804CF: ???
==31678==    by 0x117E8172: ???
==31678==    by 0x4E500AE: loader_create_instance_chain (in /home/polarina/VulkanSDK/1.0.49.0/x86_64/lib/libvulkan.so.1.0.49)
==31678==    by 0x4E52A89: vkCreateInstance (in /home/polarina/VulkanSDK/1.0.49.0/x86_64/lib/libvulkan.so.1.0.49)
==31678==    by 0x108831: main (vkleak.c:8)
==31678== 
==31678== 688 (592 direct, 96 indirect) bytes in 1 blocks are definitely lost in loss record 906 of 918
==31678==    at 0x4C2C21F: operator new(unsigned long) (vg_replace_malloc.c:334)
==31678==    by 0x10431EAB: ???
==31678==    by 0x10433A0F: ???
==31678==    by 0x10969207: ???
==31678==    by 0x10CEE3D9: ???
==31678==    by 0x11129649: ???
==31678==    by 0x1148049F: ???
==31678==    by 0x117E8172: ???
==31678==    by 0x4E500AE: loader_create_instance_chain (in /home/polarina/VulkanSDK/1.0.49.0/x86_64/lib/libvulkan.so.1.0.49)
==31678==    by 0x4E52A89: vkCreateInstance (in /home/polarina/VulkanSDK/1.0.49.0/x86_64/lib/libvulkan.so.1.0.49)
==31678==    by 0x108831: main (vkleak.c:8)
==31678== 
==31678== 848 (664 direct, 184 indirect) bytes in 1 blocks are definitely lost in loss record 907 of 918
==31678==    at 0x4C2C21F: operator new(unsigned long) (vg_replace_malloc.c:334)
==31678==    by 0x10CEE06B: ???
==31678==    by 0x10CEE3FB: ???
==31678==    by 0x11129649: ???
==31678==    by 0x1148049F: ???
==31678==    by 0x117E8172: ???
==31678==    by 0x4E500AE: loader_create_instance_chain (in /home/polarina/VulkanSDK/1.0.49.0/x86_64/lib/libvulkan.so.1.0.49)
==31678==    by 0x4E52A89: vkCreateInstance (in /home/polarina/VulkanSDK/1.0.49.0/x86_64/lib/libvulkan.so.1.0.49)
==31678==    by 0x108831: main (vkleak.c:8)
==31678== 
==31678== 904 (368 direct, 536 indirect) bytes in 1 blocks are definitely lost in loss record 908 of 918
==31678==    at 0x4C2C21F: operator new(unsigned long) (vg_replace_malloc.c:334)
==31678==    by 0x10963AD3: ???
==31678==    by 0x1096922A: ???
==31678==    by 0x10CEE3D9: ???
==31678==    by 0x11129649: ???
==31678==    by 0x1148049F: ???
==31678==    by 0x117E8172: ???
==31678==    by 0x4E500AE: loader_create_instance_chain (in /home/polarina/VulkanSDK/1.0.49.0/x86_64/lib/libvulkan.so.1.0.49)
==31678==    by 0x4E52A89: vkCreateInstance (in /home/polarina/VulkanSDK/1.0.49.0/x86_64/lib/libvulkan.so.1.0.49)
==31678==    by 0x108831: main (vkleak.c:8)
==31678== 
==31678== 6,824 (1,880 direct, 4,944 indirect) bytes in 1 blocks are definitely lost in loss record 913 of 918
==31678==    at 0x4C2C21F: operator new(unsigned long) (vg_replace_malloc.c:334)
==31678==    by 0x1111528B: ???
==31678==    by 0x1112966B: ???
==31678==    by 0x1148049F: ???
==31678==    by 0x117E8172: ???
==31678==    by 0x4E500AE: loader_create_instance_chain (in /home/polarina/VulkanSDK/1.0.49.0/x86_64/lib/libvulkan.so.1.0.49)
==31678==    by 0x4E52A89: vkCreateInstance (in /home/polarina/VulkanSDK/1.0.49.0/x86_64/lib/libvulkan.so.1.0.49)
==31678==    by 0x108831: main (vkleak.c:8)
==31678== 
==31678== 7,768 (4,944 direct, 2,824 indirect) bytes in 1 blocks are definitely lost in loss record 914 of 918
==31678==    at 0x4C2C21F: operator new(unsigned long) (vg_replace_malloc.c:334)
==31678==    by 0x117D266B: ???
==31678==    by 0x117E819B: ???
==31678==    by 0x4E500AE: loader_create_instance_chain (in /home/polarina/VulkanSDK/1.0.49.0/x86_64/lib/libvulkan.so.1.0.49)
==31678==    by 0x4E52A89: vkCreateInstance (in /home/polarina/VulkanSDK/1.0.49.0/x86_64/lib/libvulkan.so.1.0.49)
==31678==    by 0x108831: main (vkleak.c:8)
==31678== 
==31678== LEAK SUMMARY:
==31678==    definitely lost: 9,448 bytes in 7 blocks
==31678==    indirectly lost: 8,592 bytes in 85 blocks
==31678==      possibly lost: 0 bytes in 0 blocks
==31678==    still reachable: 158,015 bytes in 902 blocks
==31678==         suppressed: 0 bytes in 0 blocks
==31678== Reachable blocks (those to which a pointer was found) are not shown.
==31678== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==31678== 
==31678== For counts of detected and suppressed errors, rerun with: -v
==31678== ERROR SUMMARY: 7 errors from 7 contexts (suppressed: 0 from 0)
```

**Comment by lenny-lunarg** [comment]
I'm reassigning this to @jeremy-lunarg  because it looks like the leak is happening in the layers. I replaced `vulkaninfo.c` with your code since I didn't feel like compiling a separate program and then ran the program with

```
LD_PRELOAD="layers/libVkLayer_threading.so:layers/libVkLayer_parameter_validation.so:layers/libVkLayer_object_tracker.so:layers/libVkLayer_core_validation.so:layers/libVkLayer_swapchain.so:layers/libVkLayer_unique_objects.so" valgrind --leak-check=full demos/vulkaninfo
```

This gives the output:

```
==22906== HEAP SUMMARY:
==22906==     in use at exit: 84,768 bytes in 22 blocks
==22906==   total heap usage: 46,813 allocs, 46,791 frees, 608,270,911 bytes allocated
==22906== 
==22906== 200 (24 direct, 176 indirect) bytes in 1 blocks are definitely lost in loss record 13 of 22
==22906==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==22906==    by 0x9014323: ???
==22906==    by 0x9011FDA: ???
==22906==    by 0x838CB65: ???
==22906==    by 0x8390C69: ???
==22906==    by 0x8384728: ???
==22906==    by 0x8384A74: ???
==22906==    by 0x8384AC3: ???
==22906==    by 0x8372B85: ???
==22906==    by 0x89C380D: ???
==22906==    by 0x89C3B9F: ???
==22906==    by 0x67932C7: terminator_CreateInstance (loader.c:4690)
==22906== 
==22906== 440 (416 direct, 24 indirect) bytes in 1 blocks are definitely lost in loss record 14 of 22
==22906==    at 0x4C2E0EF: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==22906==    by 0x60B726D: layer_data* GetLayerDataPtr<layer_data>(void*, std::unordered_map<void*, layer_data*, std::hash<void*>, std::equal_to<void*>, std::allocator<std::pair<void* const, layer_data*> > >&) (vk_layer_data.h:36)
==22906==    by 0x60B1CEB: swapchain::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (swapchain.cpp:130)
==22906==    by 0x596F8DC: core_validation::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (core_validation.cpp:3395)
==22906==    by 0x552E95B: object_tracker::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (object_tracker.cpp:3429)
==22906==    by 0x517BB24: parameter_validation::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (parameter_validation.cpp:1319)
==22906==    by 0x4E5D8EA: threading::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (threading.cpp:62)
==22906==    by 0x6792059: loader_create_instance_chain (loader.c:4246)
==22906==    by 0x6797413: vkCreateInstance (trampoline.c:330)
==22906==    by 0x4008D3: main (vulkaninfo.c:9)
==22906== 
==22906== 456 bytes in 1 blocks are definitely lost in loss record 15 of 22
==22906==    at 0x4C2E0EF: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==22906==    by 0x556545F: initInstanceTable(VkInstance_T*, void (*(*)(VkInstance_T*, char const*))(), std::unordered_map<void*, VkLayerInstanceDispatchTable_*, std::hash<void*>, std::equal_to<void*>, std::allocator<std::pair<void* const, VkLayerInstanceDispatchTable_*> > >&) (vk_layer_table.cpp:147)
==22906==    by 0x552E9B9: object_tracker::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (object_tracker.cpp:3436)
==22906==    by 0x517BB24: parameter_validation::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (parameter_validation.cpp:1319)
==22906==    by 0x4E5D8EA: threading::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (threading.cpp:62)
==22906==    by 0x6792059: loader_create_instance_chain (loader.c:4246)
==22906==    by 0x6797413: vkCreateInstance (trampoline.c:330)
==22906==    by 0x4008D3: main (vulkaninfo.c:9)
==22906== 
==22906== 552 (544 direct, 8 indirect) bytes in 1 blocks are definitely lost in loss record 16 of 22
==22906==    at 0x4C2E0EF: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==22906==    by 0x51A67C5: parameter_validation::instance_layer_data* GetLayerDataPtr<parameter_validation::instance_layer_data>(void*, std::unordered_map<void*, parameter_validation::instance_layer_data*, std::hash<void*>, std::equal_to<void*>, std::allocator<std::pair<void* const, parameter_validation::instance_layer_data*> > >&) (vk_layer_data.h:36)
==22906==    by 0x517BB58: parameter_validation::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (parameter_validation.cpp:1322)
==22906==    by 0x4E5D8EA: threading::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (threading.cpp:62)
==22906==    by 0x6792059: loader_create_instance_chain (loader.c:4246)
==22906==    by 0x6797413: vkCreateInstance (trampoline.c:330)
==22906==    by 0x4008D3: main (vulkaninfo.c:9)
==22906== 
==22906== 608 (600 direct, 8 indirect) bytes in 1 blocks are definitely lost in loss record 17 of 22
==22906==    at 0x4C2E0EF: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==22906==    by 0x644BAFA: unique_objects::instance_layer_data* GetLayerDataPtr<unique_objects::instance_layer_data>(void*, std::unordered_map<void*, unique_objects::instance_layer_data*, std::hash<void*>, std::equal_to<void*>, std::allocator<std::pair<void* const, unique_objects::instance_layer_data*> > >&) (vk_layer_data.h:36)
==22906==    by 0x6445B23: unique_objects::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (unique_objects.cpp:128)
==22906==    by 0x60B1CBC: swapchain::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (swapchain.cpp:125)
==22906==    by 0x596F8DC: core_validation::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (core_validation.cpp:3395)
==22906==    by 0x552E95B: object_tracker::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (object_tracker.cpp:3429)
==22906==    by 0x517BB24: parameter_validation::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (parameter_validation.cpp:1319)
==22906==    by 0x4E5D8EA: threading::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (threading.cpp:62)
==22906==    by 0x6792059: loader_create_instance_chain (loader.c:4246)
==22906==    by 0x6797413: vkCreateInstance (trampoline.c:330)
==22906==    by 0x4008D3: main (vulkaninfo.c:9)
==22906== 
==22906== 688 (680 direct, 8 indirect) bytes in 1 blocks are definitely lost in loss record 18 of 22
==22906==    at 0x4C2E0EF: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==22906==    by 0x59AA8E0: core_validation::instance_layer_data* GetLayerDataPtr<core_validation::instance_layer_data>(void*, std::unordered_map<void*, core_validation::instance_layer_data*, std::hash<void*>, std::equal_to<void*>, std::allocator<std::pair<void* const, core_validation::instance_layer_data*> > >&) (vk_layer_data.h:36)
==22906==    by 0x596F90B: core_validation::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (core_validation.cpp:3398)
==22906==    by 0x552E95B: object_tracker::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (object_tracker.cpp:3429)
==22906==    by 0x517BB24: parameter_validation::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (parameter_validation.cpp:1319)
==22906==    by 0x4E5D8EA: threading::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (threading.cpp:62)
==22906==    by 0x6792059: loader_create_instance_chain (loader.c:4246)
==22906==    by 0x6797413: vkCreateInstance (trampoline.c:330)
==22906==    by 0x4008D3: main (vulkaninfo.c:9)
==22906== 
==22906== 3,880 (1,896 direct, 1,984 indirect) bytes in 1 blocks are definitely lost in loss record 20 of 22
==22906==    at 0x4C2E0EF: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==22906==    by 0x5552E7F: object_tracker::layer_data* GetLayerDataPtr<object_tracker::layer_data>(void*, std::unordered_map<void*, object_tracker::layer_data*, std::hash<void*>, std::equal_to<void*>, std::allocator<std::pair<void* const, object_tracker::layer_data*> > >&) (vk_layer_data.h:36)
==22906==    by 0x552E98A: object_tracker::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (object_tracker.cpp:3434)
==22906==    by 0x517BB24: parameter_validation::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (parameter_validation.cpp:1319)
==22906==    by 0x4E5D8EA: threading::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (threading.cpp:62)
==22906==    by 0x6792059: loader_create_instance_chain (loader.c:4246)
==22906==    by 0x6797413: vkCreateInstance (trampoline.c:330)
==22906==    by 0x4008D3: main (vulkaninfo.c:9)
==22906== 
==22906== 5,208 (5,200 direct, 8 indirect) bytes in 1 blocks are definitely lost in loss record 21 of 22
==22906==    at 0x4C2E0EF: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==22906==    by 0x4E79441: layer_data* GetLayerDataPtr<layer_data>(void*, std::unordered_map<void*, layer_data*, std::hash<void*>, std::equal_to<void*>, std::allocator<std::pair<void* const, layer_data*> > >&) (vk_layer_data.h:36)
==22906==    by 0x4E5D919: threading::CreateInstance(VkInstanceCreateInfo const*, VkAllocationCallbacks const*, VkInstance_T**) (threading.cpp:65)
==22906==    by 0x6792059: loader_create_instance_chain (loader.c:4246)
==22906==    by 0x6797413: vkCreateInstance (trampoline.c:330)
==22906==    by 0x4008D3: main (vulkaninfo.c:9)
==22906== 
==22906== LEAK SUMMARY:
==22906==    definitely lost: 9,816 bytes in 8 blocks
==22906==    indirectly lost: 2,216 bytes in 12 blocks
==22906==      possibly lost: 0 bytes in 0 blocks
==22906==    still reachable: 72,736 bytes in 2 blocks
==22906==         suppressed: 0 bytes in 0 blocks
==22906== Reachable blocks (those to which a pointer was found) are not shown.
==22906== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==22906== 
==22906== For counts of detected and suppressed errors, rerun with: -v
==22906== Use --track-origins=yes to see where uninitialised values come from
==22906== ERROR SUMMARY: 3829 errors from 16 contexts (suppressed: 0 from 0)
```

The first leak appears to be a driver issue, which is why I see it and this wasn't reported by @polarina. But the leaks after that are coming from a line in `vk_layer_data.h` which allocates a new pointer that then gets returned. Either the pointer isn't getting freed or something weird is happening that is throwing off valgrind. @jeremy-lunarg  knows the layers better than me, so he can debug this.

**Comment by polarina** [comment]
Think this issue should be closed, the leaks reported here were fixed when #1841 got merged. :)

**Comment by mark-lunarg** [comment]
@polarina, Thanks!
